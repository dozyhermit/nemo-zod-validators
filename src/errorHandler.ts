import { z } from 'zod';
import type { ZodError } from './types';

export const DEFAULT_ERROR_MESSAGE = 'Validation has failed';

// TECHNICAL DEBT: remove .flatten(), use z.flattenError instead
const getErrors = (
  error: ZodError['error'],
  key: keyof z.core.$ZodFlattenedError<string, unknown>
) => {
  if (typeof error?.flatten !== 'function') {
    return undefined;
  }

  if (Object.values(error?.flatten()?.[key] || {}).length <= 0) {
    return undefined;
  }

  return error.flatten()[key];
};

export const errorHandlerWithSchema = ({ error }: ZodError) =>
  Response.json(
    {
      message: getErrors(error, 'fieldErrors') || DEFAULT_ERROR_MESSAGE,
    },
    { status: 422 }
  );

export const errorHandlerWithSchemaFormErrors = ({ error }: ZodError) =>
  Response.json(
    {
      message: getErrors(error, 'formErrors') || DEFAULT_ERROR_MESSAGE,
    },
    { status: 422 }
  );

export const errorHandlerWithTransform = () =>
  Response.json(
    { message: 'Invalid request', success: false },
    {
      status: 400,
    }
  );
