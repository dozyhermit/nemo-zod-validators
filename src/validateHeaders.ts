import { NemoEvent } from '@rescale/nemo';
import { NextRequest } from 'next/server';
import type { ValidateWithSchema } from './types';
import { validateGeneric } from './validateGeneric';

type ValidateHeaders = ValidateWithSchema;

/**
 * Validates the `request.headers` tuple inside a `NextRequest` request using `zod.safeParse`.
 *
 * @param {T} schema The schema doing the validating
 * @param {Function} errorHandlerCustom A custom error function
 */
export const validateHeaders = <T>({
  schema,
  errorHandlerCustom,
}: ValidateHeaders) => {
  return async (request: NextRequest, _event: NemoEvent) =>
    validateGeneric<T>({
      data: Object.fromEntries(request?.headers || []) as T,
      schema,
      errorHandlerCustom,
    });
};
