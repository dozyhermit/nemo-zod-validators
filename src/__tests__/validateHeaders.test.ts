import { validateHeaders } from '../validateHeaders';
import { errorHandler } from './mocks/errorHandler';
import {
  type SchemaType,
  schema,
  schemaWithIntentionallyBrokenSafeParse,
} from './mocks/schema';

describe('validateHeaders', () => {
  test('passes validation with correct data', async () => {
    const action = validateHeaders<SchemaType>({ schema });

    const result = await action({
      request: {
        // @ts-expect-error type inconsistencies due to mocking headers
        headers: Object.entries({
          Number: 123,
          String: 'Hello',
          AdvancedString: '123456',
        }),
      },
    });

    expect(result.status).toBe(200);
  });

  test('triggers validation error with incorrect data', async () => {
    const action = validateHeaders<SchemaType>({ schema });

    const result = await action({
      request: {
        // @ts-expect-error type inconsistencies due to mocking headers
        headers: Object.entries({
          Number: 123,
          String: 'Hello',
          AdvancedString: 'World',
        }),
      },
    });

    expect(await result.json()).toMatchSnapshot();
  });

  test('triggers validation error with incorrect data and uses DEFAULT_ERROR_MESSAGE', async () => {
    const action = validateHeaders<SchemaType>({
      // @ts-expect-error
      schema: schemaWithIntentionallyBrokenSafeParse,
    });

    const result = await action({
      request: {
        // @ts-expect-error type inconsistencies due to mocking body
        body: Object.entries({
          Number: 123,
          String: 'Hello',
          AdvancedString: 'World',
        }),
      },
    });

    expect(await result.json()).toMatchSnapshot();
  });

  describe('with custom error handler', () => {
    test('passes validation with correct data', async () => {
      const action = validateHeaders<SchemaType>({ schema, errorHandler });

      const result = await action({
        request: {
          // @ts-expect-error type inconsistencies due to mocking headers
          headers: Object.entries({
            Number: 123,
            String: 'Hello',
            AdvancedString: '123456',
          }),
        },
      });

      expect(result.status).toBe(200);
    });

    test('triggers validation error with incorrect data', async () => {
      const action = validateHeaders<SchemaType>({ schema, errorHandler });

      const result = await action({
        request: {
          // @ts-expect-error type inconsistencies due to mocking headers
          headers: Object.entries({
            Number: 123,
            String: 'Hello',
            AdvancedString: 'World',
          }),
        },
      });

      expect(await result.json()).toMatchSnapshot();
    });
  });
});
