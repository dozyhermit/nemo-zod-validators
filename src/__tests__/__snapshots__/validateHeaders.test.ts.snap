// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`validateHeaders triggers validation error with incorrect data 1`] = `
{
  "message": {
    "AdvancedString": [
      "Invalid AdvancedString regex",
    ],
  },
}
`;

exports[`validateHeaders triggers validation error with incorrect data and uses DEFAULT_ERROR_MESSAGE 1`] = `
{
  "message": "Validation has failed",
}
`;

exports[`validateHeaders with custom error handler triggers validation error with incorrect data 1`] = `
{
  "message": "Invalid validation!",
}
`;
